pkg/intel/me/me.go

```diff
@@ -0,0 +1,280 @@
+	FitcMinor  uint16
+	FitcHotfix uint16
+	FitcBuild  uint16
+}
+
+func (h FlashPartitionTableHeader) String() string {
+	var b strings.Builder
+
+	b.WriteString("Flash partition table:\n")
+	fmt.Fprintf(&b, " Entries            : %d\n", h.NumFptEntries)
+	fmt.Fprintf(&b, " HeaderVersion      : 0x%x\n", h.HeaderVersion)
+	fmt.Fprintf(&b, " EntryVersion       : 0x%x\n", h.EntryVersion)
+	fmt.Fprintf(&b, " HeaderLength       : 0x%x\n", h.HeaderLength)
+	fmt.Fprintf(&b, " HeaderChecksum     : 0x%x\n", h.HeaderChecksum)
+	fmt.Fprintf(&b, " TicksToAdd         : 0x%x\n", h.TicksToAdd)
+	fmt.Fprintf(&b, " TokensToAdd        : 0x%x\n", h.TokensToAdd)
+	fmt.Fprintf(&b, " UMASizeOrReserved  : 0x%x\n", h.UMASizeOrReserved)
+	fmt.Fprintf(&b, " FlashLayoutOrFlags : 0x%x\n", h.FlashLayoutOrFlags)
+	fmt.Fprintf(&b, " Fitc Version       : %d.%d.%d.%d\n", h.FitcMajor, h.FitcMinor, h.FitcHotfix, h.FitcBuild)
+
+	return b.String()
+}
+
+// FlashPartitionTableEntry describes information of a flash partition table entry.
+type FlashPartitionTableEntry struct {
+	Name           [4]uint8
+	Owner          [4]uint8
+	Offset         uint32
+	Length         uint32
+	StartTokens    uint32
+	MaxTokens      uint32
+	ScratchSectors uint32
+	Flags          uint32
+}
+
+func (e FlashPartitionTableEntry) String() string {
+
+	var b strings.Builder
+
+	b.WriteString("Flash partition entry:\n")
+	fmt.Fprintf(&b, " Name          : %s\n", []byte{e.Name[0], e.Name[1], e.Name[2], e.Name[3]})
+	fmt.Fprintf(&b, " Owner         : %s\n", []byte{e.Owner[0], e.Owner[1], e.Owner[2], e.Owner[3]})
+	fmt.Fprintf(&b, " Offset        : 0x%x\n", e.Offset)
+	fmt.Fprintf(&b, " Length        : 0x%x\n", e.Length)
+	fmt.Fprintf(&b, " StartTokens   : 0x%x\n", e.StartTokens)
+	fmt.Fprintf(&b, " MaxTokens     : 0x%x\n", e.MaxTokens)
+	fmt.Fprintf(&b, " ScratchSectors: 0x%x\n", e.ScratchSectors)
+	fmt.Fprintf(&b, " Flags         : 0x%x\n", e.Flags)
+
+	if e.Flags>>24 == 0xff {
+		b.WriteString(" Valid         : No\n")
+	} else {
+		b.WriteString(" Valid         : yes\n")
+	}
+	if e.Flags&1 > 0 {
+		b.WriteString(" Partition     : Data\n")
+	} else {
+		b.WriteString(" Partition     : Code\n")
+	}
+
+	return b.String()
+}
+
+// IntelME abstracts the ME/CSME/SPS firmware found on intel platforms
+type IntelME struct {
+	hdr        FlashPartitionTableHeader
+	legacyhdr  LegacyFlashPartitionTableHeader
+	legacy     bool
+	partitions []FlashPartitionTableEntry
+	image      []byte
+	// Offset in image to $FPT
+	fptoffset uint32
+}
+
+// ParseIntelFirmware parses the Intel firmware image by uefi.Firmware interface`
+func ParseIntelFirmware(firmware uefi.Firmware) (*IntelME, error) {
+	uefi, err := ParseIntelFirmwareBytes(firmware.Buf())
+	if err != nil {
+		return nil, fmt.Errorf("unable to get the content of file: %v", err)
+	}
+
+	return uefi, nil
+}
+
+// ParseIntelFirmwareBytes parses the Intel firmware image from bytes
+func ParseIntelFirmwareBytes(imageBytes []byte) (me *IntelME, err error) {
```

Comment:
Can a unit test be added for this function?

Reply:
Yes it could but I rather reimplement the parsing and remove this function now than implementing a test now